### PR TITLE
Query: Add indexed property support

### LINKED ISSUE #15799: Query: Index property support

### PR DESCRIPTION
**Pull request by smitpatel**
Part of #13610
Resolves #15799

### CODE REVIEW

**Comment by smitpatel** [comment]
Ready for review.
- No tests for cosmos as Cosmos does not implement test from gears of war 😢 
- Adds basic queries we had in previous implementation, certainly we can expand more. We need to determine a strategy for query testing here. May be converting one of the property on Northwind to be indexer can give us existing tests.

File: src/EFCore/Infrastructure/MethodInfoExtensions.cs [hunk 1]
```diff
@@ -31,5 +33,17 @@ public static bool IsEFPropertyMethod([CanBeNull] this MethodInfo methodInfo)
...
+        ///     Returns true if the given method represents a indexer property access.
+        /// </summary>
+        /// <param name="methodInfo"> The method. </param>
+        /// <returns> True if the method is <see cref="EF.Property{TProperty}" />; false otherwise. </returns>
+        public static bool IsIndexerMethod([NotNull] this MethodInfo methodInfo)
+            => !methodInfo.IsStatic
+                && "get_Item".Equals(methodInfo.Name, StringComparison.Ordinal)
+                && typeof(object) == methodInfo.ReturnType
+                && methodInfo.GetParameters()?.Count() == 1
+                && typeof(string) == methodInfo.GetParameters().First().ParameterType;
```
Comment (AndriySvyryd) [inline src/EFCore/Infrastructure/MethodInfoExtensions.cs]: This won't work on VB. You need something like
```C#
var declaringType = methodInfo.DeclaringType;
if (declaringType is null)
{
    return false;
}

var indexerProperty = declaringType.FindIndexerProperty();
if (indexerProperty is null)
{
    return false;
}

return methodInfo == indexerProperty.GetMethod || methodInfo == indexerProperty.SetMethod;
```
With
```C#
private static PropertyInfo FindIndexerProperty(this Type type)
{
    var defaultPropertyAttribute = type.GetCustomAttributes<DefaultMemberAttribute>().FirstOrDefault();
    return defaultPropertyAttribute == null
        ? null
        : type.GetRuntimeProperties().FirstOrDefault(pi => pi.Name == defaultPropertyAttribute.MemberName && pi.IsIndexerProperty());
}
```

And also fix `FindIndexerPropertyInfo` in a similar way.

File: src/EFCore/Extensions/ModelExtensions.cs [hunk 2]
```diff
@@ -150,5 +151,17 @@ public static PropertyAccessMode GetPropertyAccessMode([NotNull] this IModel mod
...
+        ///     Gets a value indicating whether the given MethodInfo reprensent an indexer access.
+        /// </summary>
+        /// <param name="model"> The model to use. </param>
+        /// <param name="methodInfo"> The MethodInfo to check for. </param>
+        public static bool IsIndexerMethod([NotNull] this IModel model, [NotNull] MethodInfo methodInfo)
+             => !methodInfo.IsStatic
+                && methodInfo.IsSpecialName
+                && methodInfo.DeclaringType is Type declaringType
+                && model.AsModel().FindIndexerPropertyInfo(declaringType) is PropertyInfo indexerProperty
+                && (methodInfo == indexerProperty.GetMethod || methodInfo == indexerProperty.SetMethod);
```
Comment (AndriySvyryd) [inline src/EFCore/Extensions/ModelExtensions.cs]: So what's the perf penalty for this?

File: src/EFCore/Infrastructure/MethodInfoExtensions.cs [hunk 3]
```diff
@@ -31,5 +33,17 @@ public static bool IsEFPropertyMethod([CanBeNull] this MethodInfo methodInfo)
                 || methodInfo?.IsGenericMethod == true
                 && methodInfo.Name == nameof(EF.Property)
                 && methodInfo.DeclaringType?.FullName == _efTypeName;
+
+        /// <summary>
+        ///     Returns true if the given method represents a indexer property access.
+        /// </summary>
+        /// <param name="methodInfo"> The method. </param>
+        /// <returns> True if the method is <see cref="EF.Property{TProperty}" />; false otherwise. </returns>
+        public static bool IsIndexerMethod([NotNull] this MethodInfo methodInfo)
```
Comment (AndriySvyryd) [inline src/EFCore/Infrastructure/MethodInfoExtensions.cs]: Do we need this?

File: src/EFCore/Extensions/ModelExtensions.cs (hunk 2 above)
Comment (smitpatel) [reply src/EFCore/Extensions/ModelExtensions.cs]: The largest penalty was for GetParameters().
SqlServer.FunctionalTests
This version 2.9286 mins
get_item version 2.3472 mins
get_item short-circuit with above 3.0012 mins
With GetParmeters & above version - 17.7040 mins

File: src/EFCore/Infrastructure/MethodInfoExtensions.cs (hunk 3 above)
Comment (smitpatel) [reply src/EFCore/Infrastructure/MethodInfoExtensions.cs]: (I need to update doc for it).
May be if you don't have model and has to check. Could be YAGNI though. Your call.

File: src/EFCore/Extensions/ModelExtensions.cs (hunk 2 above)
Comment (AndriySvyryd) [reply src/EFCore/Extensions/ModelExtensions.cs]: Does removing `methodInfo.DeclaringType is Type declaringType` make any difference?

File: src/EFCore/Infrastructure/MethodInfoExtensions.cs (hunk 3 above)
Comment (AndriySvyryd) [reply src/EFCore/Infrastructure/MethodInfoExtensions.cs]: YAGNI

File: src/EFCore/Extensions/ModelExtensions.cs (hunk 2 above)
Comment (AndriySvyryd) [reply src/EFCore/Extensions/ModelExtensions.cs]: We should add a test that calls an anonymous function in the projection if we don't have one already

Comment (smitpatel) [reply src/EFCore/Extensions/ModelExtensions.cs]: how to call anonymous functions? (what are the anonymous functions? :trollface:

Comment (AndriySvyryd) [reply src/EFCore/Extensions/ModelExtensions.cs]: dbSet.Select(e => (x => x)(e))

Comment (smitpatel) [reply src/EFCore/Extensions/ModelExtensions.cs]: > Does removing methodInfo.DeclaringType is Type declaringType make any difference?

negligible.

Comment (smitpatel) [reply src/EFCore/Extensions/ModelExtensions.cs]: image

Gives error "Method name expected"

Comment (smitpatel) [reply src/EFCore/Extensions/ModelExtensions.cs]: Filed https://github.com/aspnet/EntityFrameworkCore/issues/19480

Comment (AndriySvyryd) [reply src/EFCore/Extensions/ModelExtensions.cs]: Ok, then we can remove the `methodInfo.DeclaringType is Type declaringType` line

**Pull request merged**